src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/Solution.cs

```diff
@@ -0,0 +1,167 @@
+            Descriptions = descriptions;
+            ActionArgs = actionArgs;
+            RequiresConfirmation = confirm;
+            ActionName = actionName != string.Empty ? actionName : title;
+        }
+
+        /// <summary>
+        /// Creates a Solution component. Prefer a pre-defined Solution component such as Solution.Restart.
+        /// </summary>
+        /// <param name="title">Solution title.</param>
+        /// <param name="resourceUri">The URI of the target resource on which an action will be performed.</param>
+        /// <param name="context">Determines if an internal view should be rendered.</param>
+        /// <param name="action">Denotes which action will be performed, such as a specific API call or navigation link.</param>
+        /// <param name="internalInstructions">Instructions that will be sent to the customer by support staff.</param>
+        /// <param name="descriptions">Descriptions which will be viewed as Markdown.</param>
+        /// <param name="actionArgs">Free-form JSON-serializable arguments that can be sent as the body of the
+        ///     action request if a request body is applicable.</param>
+        /// <param name="confirm">If the solution requires confirmation, it will signify a potentially dangerous action in the UI.</param>
+        /// <param name="actionName">Name of the action for display customization purposes.</param>
+        public Solution(string title, string resourceUri, OperationContext context, ActionType action,
+            string internalInstructions, IEnumerable<string> descriptions = null,
+            Dictionary<string, object> actionArgs = null, bool confirm = false, string actionName = "") :
+            this(title, resourceUri, context.IsInternalCall, action, internalInstructions, descriptions, actionArgs,
+                confirm)
+        { }
+
+        // TODO: This should pass any remaining arguments to the constructor
+        // TODO: isInternal and detectorId should be integrated in backend, not passed here
+        /// <summary>
+        /// Creates a pre-defined Solution component capable of restarting an App instance
+        /// </summary>
+        public static Solution Restart(string resourceUri, bool isInternal, string detectorId)
+        {
+            var resourceLink = UriUtilities.BuildDetectorLink(resourceUri, detectorId);
+            var instructions = $"[Go To Detector]({resourceLink})\n\n" +
+                $"{SolutionConstants.RestartInstructions}";
+
+            return new Solution("Restart Site", resourceUri, isInternal, ActionType.RestartSite, instructions,
+                new string[] { SolutionConstants.AppRestartDescription }, confirm: true, actionName: "Restart App");
+        }
+
+        /// <summary>
+        /// Creates a pre-defined Solution component capable of restarting an App instance
+        /// </summary>
+        public static Solution Restart(string resourceUri, OperationContext context, string detectorId)
```

Comment:
Does anyone know of a way to provide the OperationContext and Detector metadata to this component without relying on the detector author to provide them? They will be consistent in every Solution case, and should be taken care of without the detector author's help.

Reply:
For Detector Metadata, it is actually part of the Response object that is passed to the script.
We populate the Definition object from the Entry Point (Run method) before invoking the csx.

![image](https://user-images.githubusercontent.com/11183841/53585589-24eafd00-3b3b-11e9-8750-6cf688742cc9.png)

So you should be ale to use that in the extension method.

However, this is not true for OperationContext. So it looks like you would have to pass that explicitly.

Reply:
Would it make sense to add OperationContext in the Response object? Or since all I need is the internal/external flag, would it make sense to add that to the Response?
I imagine this isn't the only component that changes its rendering depending on internal/external view, so it's possible we could cut out a lot of extra work for detector authors on other components as well.

Reply:
So the way we differentiate between UI Views for Internal/External is in the client code actually. 
I guess I didnt understand the current flow : The backend only communicates to the frontend that there is a restart solution and the UI has selected views (and descriptions) or the Backend makes the decision for internal/external and sends different description/data to UI

Reply:
There is probably a way to remove "IsInternal" from the response object in general and then pass all the information you need to display either view (which it looks like you are doing) and have the UI know whether it is internal or not and display accordingly. The logic would need to be added in the UI but you could do something like the detector container level provides a service that has some information like this that is common for the whole detector, and then it can access it from the solution component. I also think detector-control service has access to this information already.

Reply:
@stevenernst131 I would need some more specifics on how to go that route. It does sound better in the long term to leave the Response object as simple as possible. Do you think the current implementation is fine to get this PR through?